sssom/io.py

```diff
@@ -228,3 +232,114 @@ def extract_iri(input, prefix_map) -> list:
+#     )
+#     condition_obj = predicate_filtered_df[OBJECT_ID].apply(
+#         lambda x: str(x).startswith(prefix_keys)
+#     )
+#     filtered_df = predicate_filtered_df.loc[condition_subj & condition_obj]
+
+#     new_msdf: MappingSetDataFrame = MappingSetDataFrame(
+#         df=filtered_df, prefix_map=filtered_prefix_map, metadata=msdf.metadata
+#     )
+#     return new_msdf
+
+def run_sql_query(query: str, inputs: List[str], output: TextIO):
+    """Run a SQL query over one or more SSSOM files.
+
+    Each of the N inputs is assigned a table name df1, df2, ..., dfN
+
+    Alternatively, the filenames can be used as table names - these are first stemmed
+    E.g. ~/dir/my.sssom.tsv becomes a table called 'my'
+
+    Example:
+        sssom dosql -Q "SELECT * FROM df1 WHERE confidence>0.5 ORDER BY confidence" my.sssom.tsv
+
+    Example:
+        `sssom dosql -Q "SELECT file1.*,file2.object_id AS ext_object_id, file2.object_label AS ext_object_label \
+        FROM file1 INNER JOIN file2 WHERE file1.object_id = file2.subject_id" FROM file1.sssom.tsv file2.sssom.tsv`
+
+    :param query: Query to be executed over a pandas DataFrame (msdf.df).
+    :param inputs: Input files that form the source tables for query.
+    :param output: Output.
+    """
+    n = 1
+    new_msdf = MappingSetDataFrame()
+    while len(inputs) >= n:
+        fn = inputs[n - 1]
+        msdf = parse_sssom_table(fn)
+        df = msdf.df
+        # df = parse(fn)
+        globals()[f"df{n}"] = df
+        tn = re.sub("[.].*", "", Path(fn).stem).lower()
+        globals()[tn] = df
+        n += 1
+
+    new_df = sqldf(query)
+    new_msdf.df = new_df
+    new_msdf.prefix_map = add_built_in_prefixes_to_prefix_map(msdf.prefix_map)
+    new_msdf.metadata = msdf.metadata
+    write_table(new_msdf, output)
+    return new_msdf
+
+def filter_file(input: str, output: TextIO, **kwargs):
+    """Filter a dataframe by dynamically generating queries based on user input.
+
+    e.g. sssom filter --subject_id x:% --subject_id y:% --object_id y:% --object_id z:% tests/data/basic.tsv
+
+    yields the query:
+
+    "SELECT * FROM df WHERE (subject_id LIKE 'x:%'  OR subject_id LIKE 'y:%')
+     AND (object_id LIKE 'y:%'  OR object_id LIKE 'z:%') " and displays the output.
+
+    :param input: DataFrame to be queried over.
+    :param output: Output location.
+    :param **kwargs: Filter options provided by user which generate queries (e.g.: --subject_id x:%).
+    """
+    params = {k: v for k, v in kwargs.items() if v}
+    query = "SELECT * FROM df WHERE ("
+    multiple_params = True if len(params) > 1 else False
+    for idx, (k, v) in enumerate(params.items(), start=1):
+        query += k + " LIKE '" + v[0] + "' "
+        if len(v) > 1:
+            for exp in v[1:]:
+                query += " OR "
+                query += k + " LIKE '" + exp + "') "
+        if multiple_params and idx != len(params):
+            query += " AND ("
```

Comment:
Does it make sense to throw an exception of no legal parameter is supplied? Just to be sure you don't try to query anything if there is not filter criteria?

Reply:
Absolutely! Good catch, I've updated it.